src/util/metadata/side_metadata/global.rs

```diff
@@ -946,6 +1075,194 @@ impl SideMetadataSpec {
+        let end_addr = data_addr
+            .saturating_sub(search_limit_bytes)
+            .align_down(region_bytes);
+
+        let mut cursor = start_addr;
+        while cursor > end_addr {
+            // We encounter an unmapped address. Just return None.
+            if !cursor.is_mapped() {
+                return None;
+            }
+            // If we find non-zero value, just return it.
+            if !unsafe { self.load::<T>(cursor).is_zero() } {
+                return Some(cursor);
+            }
+            cursor -= region_bytes;
+        }
+        None
+    }
+
+    #[allow(clippy::let_and_return)]
+    fn find_prev_non_zero_value_fast<T: MetadataValue>(
+        &self,
+        data_addr: Address,
+        search_limit_bytes: usize,
+    ) -> Option<Address> {
+        debug_assert!(self.uses_contiguous_side_metadata());
+
+        // Quick check if the data address is mapped at all.
+        if !data_addr.is_mapped() {
+            return None;
+        }
+        // Quick check if the current data_addr has a non zero value.
+        if !unsafe { self.load::<T>(data_addr).is_zero() } {
+            return Some(data_addr.align_down(1 << self.log_bytes_in_region));
+        }
+
+        // Figure out the start and end data address.
+        let start_addr = data_addr.saturating_sub(search_limit_bytes) + 1usize;
+        let end_addr = data_addr;
+
+        // Then figure out the start and end metadata address and bits.
+        let start_meta_addr = address_to_contiguous_meta_address(self, start_addr);
+        let start_meta_shift = meta_byte_lshift(self, start_addr);
+        let end_meta_addr = address_to_contiguous_meta_address(self, end_addr);
+        let end_meta_shift = meta_byte_lshift(self, end_addr);
+
+        // The result will be set by one of the following closures.
+        // Use Cell so it doesn't need to be mutably borrowed by the two closures which Rust will complain.
+        let res = std::cell::Cell::new(None);
+
+        // Check and find the last bit that is set. We try load words where possible, and fall back to load bytes.
+        let check_bytes_backwards = |start: Address, end: Address| -> bool {
+            use crate::util::constants::BYTES_IN_ADDRESS;
+            unsafe {
+                let mut cur = end;
+                while cur > start {
+                    // If we can check the whole word, set step to word size. Otherwise, the step is 1 (byte) and we check byte.
+                    let step = if cur.is_aligned_to(BYTES_IN_ADDRESS)
+                        && cur.align_down(BYTES_IN_ADDRESS) >= start
+                    {
+                        BYTES_IN_ADDRESS
+                    } else {
+                        1
+                    };
+                    // Move to the address so we can load from it
+                    cur -= step;
+
+                    // We have to always check if the metadata address is mapped or not.
+                    if !cur.is_mapped() {
```

Comment:
Rather than calling `is_mapped` at every `step`, we may set a sentinel (cur aligned down to page size) so that we don't need to call `is_mapped` until `cur` reaches into another page.

Reply:
Sounds good. We do mmap at chunk granularity. We can use that to only check every chunk.

Reply:
Done. This makes the benchmark 50% faster. See https://github.com/mmtk/mmtk-core/pull/1165#discussion_r1674800993.